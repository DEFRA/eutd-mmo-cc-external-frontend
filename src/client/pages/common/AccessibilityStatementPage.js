import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { getAllUserAttributes } from '../../actions/userAttributes.actions';
import { AccessibilityStatementEnglish } from './AccessibilityStatementEnglish';
import { AccessibilityStatementWelsh } from './AccessibilityStatementWelsh';

const AccessibilityStatementPage = (props) => {
  let userSelectedLanguage;
  const userAttributeLanguage = props.userAttributes.find((val) => val.name === 'language');
  
  if (userAttributeLanguage) {
    userSelectedLanguage = userAttributeLanguage.value;
  }

  return (
    userSelectedLanguage === 'cy_UK' 
      ? <AccessibilityStatementWelsh />
      : <AccessibilityStatementEnglish />
  );
};

const mapStateToProps = state => {
  return {
    userAttributes: state.userAttributes
  };
};

const loadData = store => {
  return store.dispatch(getAllUserAttributes());
};

export default {
  loadData,
  component: withRouter(connect(mapStateToProps,
    {
      getAllUserAttributes
    })(AccessibilityStatementPage))
};
