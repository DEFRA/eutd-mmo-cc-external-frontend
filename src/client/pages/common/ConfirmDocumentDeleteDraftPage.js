import React from 'react';
import {withRouter} from 'react-router';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import _ from 'lodash';

import {
  saveConfirmDocumentDelete,
  addConfirmDocumentDelete,
  dispatchApiCallFailed
} from '../../actions';
import {
  Main,
  BackLink,
  Header,
  GridRow,
  GridCol,
  MultiChoice
} from 'govuk-react';

import PageTitle from '../../components/PageTitle';
import Form from '../../components/elements/Form';
import HelpLink from '../../components/HelpLink';
import ContinueButton from '../../components/elements/ContinueButton';
import ErrorIsland from '../../components/elements/ErrorIsland';

import { scrollToErrorIsland, onHandleErrorClick} from '../utils';
import SelectRadio from '../../components/elements/SelectRadio';
import PageTemplateWrapper from '../../components/PageTemplateWrapper';
import { withTranslation } from 'react-i18next';
import i18n from '../../../../src/i18n';
class ConfirmDocumentDeleteDraftPage extends React.Component {

  componentDidMount(){
    i18n.changeLanguage();
    this.forceUpdate();
  }

  onChange = e => {
    const confirmDocumentDelete = _.cloneDeep(
      this.props.confirmDocumentDelete || {}
    );
    confirmDocumentDelete.documentDelete = e.target.value;
    this.props.addConfirmDocumentDelete(confirmDocumentDelete);
  };

  onSubmit = async e => {
    e.preventDefault();

    try {
      const {path, journey, previousUri, nextUri} = this.props.route;
      const documentNumber = this.props.match.params['documentNumber'];
      const documentDelete = await this.props.saveConfirmDocumentDelete(path, nextUri, journey, documentNumber);
      const to = (documentDelete && documentDelete === 'Yes') ? nextUri : previousUri;

      this.props.history.push(to);
    } catch (error) {
      scrollToErrorIsland();
    }
  };

  render() {
    const {documentDeleteError, errors} = this.props.errors;
    const {confirmDocumentDelete = {}, route, match, t} = this.props;
    const {path, previousUri, nextUri, journey, journeyText} = route;
    const title = t(`${journey}DeleteCatchCertificateDeleteDraftJourneytext`, {journeyText: t(journeyText)})
    const h1Text =  t(`${journey}DeleteCatchCertificateDraftDeleteConfirmationText`, {journeyText: t(journeyText)})
    const currentUri = path.replace(':documentNumber', match.params.documentNumber);
    return (
      <Main>
        {errors && (
          <ErrorIsland
            errors={errors.map(err => ({
              message: t(err.text),
              key: err.targetName
            }))}
            onHandleErrorClick={onHandleErrorClick}
          />
        )}

        <PageTitle
          title={`${!_.isEmpty(errors) ? 'Error: ' : ''}${title} - ${t(this.props.route.title)}`}
        />
        <GridRow>
            <BackLink onClick={e => { e.preventDefault(); this.props.history.push(previousUri); }} href={previousUri}>
            {t('commonBackLinkBackButtonLabel')}
              </BackLink>
        </GridRow>

        <Form action='/orchestration/api/v1/confirm-document-delete' currentUrl={currentUri} onSubmit={this.onSubmit}>
          <GridRow>
            <GridCol columnTwoThirds>
              <GridRow>
                <Header level='1'>
                  <legend>{h1Text}</legend>
                </Header>
              </GridRow>
              <MultiChoice meta={{touched: t(documentDeleteError), error: t(documentDeleteError)}} id='documentDelete'>
                <fieldset>
                  <SelectRadio id="documentDeleteYes" checked={confirmDocumentDelete.documentDelete === 'Yes'} value='Yes' name='documentDelete' inline onChange={this.onChange}>{t('Yes')}</SelectRadio>
                  <SelectRadio id="documentDeleteNo" checked={confirmDocumentDelete.documentDelete === 'No'} value='No' name='documentDelete' inline onChange={this.onChange}>{t('No')}</SelectRadio>
                </fieldset>
              </MultiChoice>
            </GridCol>
          </GridRow>
          <GridRow>
            <GridCol columnTwoThirds>
              <ContinueButton type='submit' id='continue'>{t('commonContinueButtonSaveAndContinueButton')}</ContinueButton>
            </GridCol>
          </GridRow>
          <input type="hidden" name="journey" value={journey} />
          <input type="hidden" name="currentUri" value={currentUri} />
          <input type="hidden" name="nextUri" value={nextUri} />
          <input type="hidden" name="previousUri" value={previousUri} />
        </Form>
        <HelpLink journey={route.journey} />
      </Main>
    );
  }
}

function mapStateToProps(state) {
  return {
    errors: state.errors,
    confirmDocumentDelete: state.confirmDocumentDelete
  };
}

function loadData(store) {
  if (Object.keys(this.queryParams).length !== 0) {
    store.dispatch(dispatchApiCallFailed(JSON.parse(this.queryParams.error)));
  }
}

export const component = withRouter(
  connect(
    mapStateToProps,
    {
      addConfirmDocumentDelete,
      saveConfirmDocumentDelete
    }
  )(withTranslation()(ConfirmDocumentDeleteDraftPage))
);

ConfirmDocumentDeleteDraftPage.propTypes = {
  confirmDocumentDelete: PropTypes.object.isRequired,
  errors: PropTypes.object
};

export default {
  loadData,
  component: PageTemplateWrapper(component)
};
