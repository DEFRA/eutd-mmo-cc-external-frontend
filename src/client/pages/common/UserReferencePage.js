import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { isEmpty } from 'lodash';
import { withTranslation } from 'react-i18next';
import BackToProgressLink from '../../components/BackToProgressLink';

import {
  getUserReference,
  saveUserReference,
  clearUserReference
} from '../../actions/user-reference.actions';

import {
  clearCopyDocument
} from '../../actions/copy-document.actions';

import {
  Main,
  BackLink,
  Header,
  InputField,
  GridRow,
  GridCol
} from 'govuk-react';

import { scrollToErrorIsland, onHandleErrorClick} from '../utils';

import Form from '../../components/elements/Form';
import HelpLink from '../../components/HelpLink';
import PageTitle from '../../components/PageTitle';
import SaveAsDraftButton from '../../components/SaveAsDraftButton';
import ContinueButton from '../../components/elements/ContinueButton';

import PageTemplateWrapper from '../../components/PageTemplateWrapper';
import ErrorIsland from '../../components/elements/ErrorIsland';
import NotificationBanner from '../../components/NotificationBanner';
import errorTransformer from '../../helpers/errorTransformer';

class UserReferencePage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      userReference: ''
    };
  }

  onChange = e => {
    this.setState({ userReference: e.target.value });
  }

  onSaveAsDraft = async (e) => {
    e.preventDefault();
    const { saveAsDraftUri } = this.props.route;
    await this.onSave(saveAsDraftUri);
  }

  onSave = async (nextUri) => {
    const { errors } = this.props.errors;
    const documentNumber = this.props.match.params['documentNumber'];

    try {
      await this.props.saveUserReference({ userReference: this.state.userReference.toString().trim() }, documentNumber);
      this.props.history.push(nextUri);
    } catch (e) {
      if (!isEmpty(errors)) {
        scrollToErrorIsland();
      }
    }
  }

  unauthorised() {
    this.props.history.push('/forbidden');
  }

  getCopyNotificationMsg(isVoided, documentNumber) {
    const { journeyText, journey } = this.props.route;
    const addRemoveLandingText = journey === 'catchCertificate' ? ' landing' : '';
    return isVoided ? `This draft was created by copying a document that has now been voided. You are reminded that you must not use a ${this.props.t(journeyText)} or${addRemoveLandingText} data for catches that have already been exported as this is a serious offence and may result in enforcement action being taken.` :
    `This draft was created by copying document ${documentNumber}. You are reminded that you must not use a ${this.props.t(journeyText)} or${addRemoveLandingText} data for catches that have already been exported as this is a serious offence and may result in enforcement action being taken.`;
   }

  componentDidUpdate() {
    const { reference } = this.props;

    if (reference.unauthorised === true) {
      this.unauthorised();
    }
  }

  componentDidMount = async () => {
    window.scrollTo(0, 0);

    const documentNumber = this.props.match.params['documentNumber'];

    try {
      await this.props.getUserReference(documentNumber);

      const { userReference } = this.props.reference;
      this.setState({ userReference: userReference });
    } catch (e) {
      console.error(e);
    }
  }

  componentWillUnmount() {
    this.props.clearCopyDocument();
    this.props.clearUserReference();
  }

  render = () => {
    const { route, match, confirmCopyDocument, t } = this.props;
    const { title, header, previousUri, path, nextUri, journey, journeyText } = route;
    const { documentNumber } = match.params;
    const errors = errorTransformer(this.props.reference.error);

    const saveAsDraftFormActionUrl = `/orchestration/api/v1/userReference/${journey}/saveAsDraftLink`;
    const backLinkUrl = previousUri.replace(':documentNumber', documentNumber);

    return (
      <Main>
        <PageTitle title={`${!isEmpty(errors) ? 'Error: ' : ''} ${t(header)} - ${t(title)}`} />
        {!isEmpty(errors) && errors.errors && errors.errors.length > 0 && (
          <ErrorIsland
            errors={errors.errors.map((err) => ({
              message: t(err.text),
              key: err.targetName,
            }))}
            onHandleErrorClick={onHandleErrorClick}
          />
        )}
        <GridRow>
          <GridCol>
            <BackLink onClick={e => {
              e.preventDefault();
              this.props.history.push(backLinkUrl);
            }
            }
              href={backLinkUrl}>
              {t('commonBackLinkBackButtonLabel')}
            </BackLink>
            {confirmCopyDocument.copyDocumentAcknowledged && (
              <NotificationBanner
                header={t('commonImportant')}
                messages={[this.getCopyNotificationMsg(confirmCopyDocument.voidDocumentConfirm, confirmCopyDocument.documentNumber)]} />
            )}
           <Header>
             {t(header)}
            </Header>
          </GridCol>
        </GridRow>
        <Form
          action="/orchestration/api/v1/userReference"
          currentUrl={path.replace(':documentNumber', documentNumber)}
          nextUrl={nextUri.replace(':documentNumber', documentNumber)}
          documentNumber={documentNumber}
          onSubmit={e => e.preventDefault()}>
          <input type="hidden" name="journey" value={journey} />
          <input type="hidden" name="currentUri" value={path} />
          <GridRow>
            <GridCol columnTwoThirds>
              <InputField
                meta={{ error: t(errors.userReferenceError), touched: true }}
                htmlFor={'userReference'}
                input={{
                  autoComplete: 'off',
                  id: 'userReference',
                  className: 'formControl',
                  name: 'userReference',
                  value: this.state.userReference,
                  onChange: e => this.onChange(e),
                  onBlur: e => this.onChange(e)
                }}
                hint={t('commonAddYourReferenceHint', {journeyText: t(journeyText)})}>{t('commonAddYourReferenceOptionalText')}
                </InputField>
            </GridCol>
          </GridRow>
          <GridRow>
            <SaveAsDraftButton formactionUrl={saveAsDraftFormActionUrl} onClick={this.onSaveAsDraft} />
            <ContinueButton id="continue" onClick={() => this.onSave(nextUri.replace(':documentNumber', documentNumber))}>
              {t('commonContinueButtonSaveAndContinueButton')}
            </ContinueButton>
          </GridRow>
        </Form>
        <BackToProgressLink
          progressUri={previousUri}
          documentNumber={documentNumber}
        />
        <HelpLink journey={journey} />
      </Main>
    );
  }
}

function mapStateToProps(state) {
  return {
    confirmCopyDocument: state.confirmCopyDocument,
    reference: state.reference,
    errors: state.errors
  };
}

function loadData(store) {
  return store.dispatch(getUserReference(this.documentNumber));
}

export const component = withRouter(
  connect(mapStateToProps, {
    getUserReference,
    saveUserReference,
    clearUserReference,
    clearCopyDocument
  })(withTranslation() (UserReferencePage)));

UserReferencePage.propTypes = {
  confirmCopyDocument: PropTypes.object,
  reference: PropTypes.object,
  errors: PropTypes.object
};

export default {
  loadData,
  component: PageTemplateWrapper(component)
};
