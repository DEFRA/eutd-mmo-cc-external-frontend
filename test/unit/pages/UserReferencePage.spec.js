import _ from 'lodash';
import { mount } from 'enzyme';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter, Router, Route } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import thunk from 'redux-thunk';
import { component as UserReferencePage } from '../../../src/client/pages/common/UserReferencePage';
import userReference from '../../../src/client/pages/common/UserReferencePage';
import {
  getUserReference,
  clearUserReference,
} from '../../../src/client/actions/user-reference.actions';
import { BackLink } from 'govuk-react';
import { render } from '@testing-library/react';

jest.mock('../../../src/client/actions/user-reference.actions');

beforeEach(() => {
  getUserReference.mockReturnValue({ type: 'GET_USER_REFERENCE' });
  clearUserReference.mockReturnValue({ type: 'CLEAR_USER_REFERENCE' });
});

describe('Add user reference to document', () => {

  let wrapper;

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  let mockPush;

  const props = {
    route: {
      title: 'Create a UK catch certificate - GOV.UK',
      header: 'Add your reference for this export',
      path: '/create-catch-certificate/:documentNumber/add-your-reference',
      previousUri: '/create-catch-certificate/:documentNumber/progress',
      nextUri: '/create-catch-certificate/:documentNumber/add-exporter-details',
      saveAsDraftUri: '/create-catch-certificate/catch-certificates',
      journey: 'catchCertificate',
      journeyText: 'catchCertificate'
    },
    match: {
      params: {
        documentNumber: 'GBR-23423-4234234'
      }
    },
    t: jest.fn()
  };

  const documentNumber = 'GBR-23423-4234234';

  beforeEach(() => {
    const store = mockStore({
      reference: { userReference: 'My Reference' },
      confirmCopyDocument: { copyDocumentAcknowledged: false, voidDocumentConfirm: false },
      errors: {},
      t: jest.fn()
    });

    window.scrollTo = jest.fn();

    const history = createMemoryHistory({ initialEntries: [`/create-catch-certificate/${documentNumber}/add-your-reference`] });

    mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <Route path='/create-catch-certificate/:documentNumber/add-your-reference'>
            <UserReferencePage {...props}/>
          </Route>
        </Router>
      </Provider>
    );
  });

  it('should render user reference text input', () => {
    expect(wrapper.find('input#userReference').exists()).toBeTruthy();
  });

  it('should contain back button that goes to the progress page', () => {
    expect(
      wrapper.find('a[href$="/create-catch-certificate/GBR-23423-4234234/progress"]').exists()
    ).toBeTruthy();
  });

  it('will redirect to the progress uri when back link is clicked', () => {
    const previousUri = props.route.previousUri;
    const documentNumber = props.match.params.documentNumber;

    wrapper.find(BackLink).simulate('click');

    expect(mockPush).toHaveBeenCalledWith(
      previousUri.replace(':documentNumber', documentNumber)
    );
  });

  it('should handle submit event', () => {
    wrapper.find('form').simulate(
      'submit',
      {preventDefault() {}}
    );
  });

  it('should handle on change events', () => {
    wrapper.find('input[name="userReference"]').simulate('change', {target: {name: 'userReference',  value: 'My Reference'}});
  });

  it('should handle on blur events', () => {
    wrapper.find('input[name="userReference"]').simulate('blur', { target: { name: 'userReference', value: 'My Reference 1' } });
    expect(wrapper.find('input[name="userReference"]').instance().value).toEqual('My Reference 1');
  });

  it('should handle continue button click event', () => {
    wrapper.setState({selected: '/create-catch-certificate/catch-certificates'});
    wrapper.find('button#continue').simulate('click');
});

  it('should handle save as draft event', () => {
    wrapper.find('button#saveAsDraft').simulate('click');
  });

  it('should have a label with a userReference for attribute', () => {
    expect(wrapper.find('label').props()['htmlFor']).toBeDefined();
    expect(wrapper.find('label').props()['htmlFor']).toBe('userReference');
  });

  it('should not display copy document notification', () => {
    expect(wrapper.find('NotificationBanner').exists()).toBeFalsy();
  });

  it('should have the correct form parameters', () => {
    const currentUri = props.route.path.replace(':documentNumber', documentNumber);
    const nextUri = props.route.nextUri.replace(':documentNumber', documentNumber);
    const form = wrapper.find('Form');

    expect(form.prop('action')).toBe('/orchestration/api/v1/userReference');
    expect(form.prop('nextUrl')).toBe(nextUri);
    expect(form.prop('currentUrl')).toBe(currentUri);
    expect(form.prop('documentNumber')).toBe(documentNumber);
  });
});

describe('When redirected from copy confirm document page', () => {
  let wrapper;

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  const storeDefaults = {
    reference: { userReference: 'My Reference' },
    confirmCopyDocument: { copyDocumentAcknowledged: true, voidDocumentConfirm: false },
    errors: {},
    t: jest.fn()
  };

  const props = {
    history: [],
    route: {
      title: 'Create a UK catch certificate - GOV.UK',
      header: 'Add your reference for this export',
      path: '/create-catch-certificate/:documentNumber/add-your-reference',
      previousUri: '/create-catch-certificate/catch-certificates',
      nextUri: '/create-catch-certificate/:documentNumber/add-exporter-details',
      saveAsDraftUri: '/create-catch-certificate/catch-certificates',
      journey: 'catchCertificate',
      journeyText: 'catch certificate'
    },
    match : {
      params : {
        documentNumber: 'GBR-23423-4234234'
      }
    },
    t: jest.fn()
  };


  beforeEach(() => {
    window.scrollTo = jest.fn();
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it('should display copy document notification for catch certificate', () => {
    wrapper = mount(
      <Provider store={mockStore(storeDefaults)}>
        <MemoryRouter>
          <UserReferencePage {...props}/>
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('NotificationBanner').exists()).toBeTruthy();
    expect(wrapper.find('NotificationBanner').prop('messages')[0]).toBe('This draft was created by copying document undefined. You are reminded that you must not use a catch certificate or landing data for catches that have already been exported as this is a serious offence and may result in enforcement action being taken.');
  });

  it('should display copy document notification with a VOID message for catch certificate', () => {
    const store = _.cloneDeep(storeDefaults);

    store.confirmCopyDocument.voidDocumentConfirm = true;

    wrapper = mount(
      <Provider store={mockStore(store)}>
        <MemoryRouter>
          <UserReferencePage {...props}/>
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('NotificationBanner').exists()).toBeTruthy();
    expect(wrapper.find('NotificationBanner').prop('messages')[0]).toBe('This draft was created by copying a document that has now been voided. You are reminded that you must not use a catch certificate or landing data for catches that have already been exported as this is a serious offence and may result in enforcement action being taken.');
  });

  it('should not display copy document notification with a VOID message', () => {
    const store = _.cloneDeep(storeDefaults);

    store.confirmCopyDocument.copyDocumentAcknowledged = false;
    store.confirmCopyDocument.voidDocumentConfirm = true;

    wrapper = mount(
      <Provider store={mockStore(store)}>
        <MemoryRouter>
          <UserReferencePage {...props}/>
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('NotificationBanner').exists()).toBeFalsy();
  });

});

describe('When redirected from copy comfirm document page for a processing statement', () => {
  let wrapper;

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  const storeDefaults = {
    reference: { userReference: 'My Reference' },
    confirmCopyDocument: { copyDocumentAcknowledged: true, voidDocumentConfirm: false },
    errors: {}
  };

  const props = {
    history: [],
    route: {
      title: 'Create a UK catch certificate - GOV.UK',
      header: 'Add your reference for this export',
      path: '/create-catch-certificate/:documentNumber/add-your-reference',
      previousUri: '/create-catch-certificate/catch-certificates',
      nextUri: '/create-catch-certificate/:documentNumber/add-exporter-details',
      saveAsDraftUri: '/create-catch-certificate/catch-certificates',
      journey: 'processingStatement',
      journeyText: 'processing statement'
    },
    match : {
      params : {
        documentNumber: 'GBR-23423-4234234'
      }
    }
  };

  beforeEach(() => {
    window.scrollTo = jest.fn();
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it('should display copy document notification for processing statement', () => {
    wrapper = mount(
      <Provider store={mockStore(storeDefaults)}>
        <MemoryRouter>
          <UserReferencePage {...props}/>
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('NotificationBanner').exists()).toBeTruthy();
    expect(wrapper.find('NotificationBanner').prop('messages')[0]).toBe('This draft was created by copying document undefined. You are reminded that you must not use a processing statement or data for catches that have already been exported as this is a serious offence and may result in enforcement action being taken.');
  });

  it('should display copy document notification with a VOID message for processing statement', () => {
    const store = _.cloneDeep(storeDefaults);

    store.confirmCopyDocument.voidDocumentConfirm = true;

    wrapper = mount(
      <Provider store={mockStore(store)}>
        <MemoryRouter>
          <UserReferencePage {...props}/>
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('NotificationBanner').exists()).toBeTruthy();
    expect(wrapper.find('NotificationBanner').prop('messages')[0]).toBe('This draft was created by copying a document that has now been voided. You are reminded that you must not use a processing statement or data for catches that have already been exported as this is a serious offence and may result in enforcement action being taken.');
  });

  it('should take a snapshot of the whole page', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });
});

describe('loadData', () => {
  const store = {
    dispatch: () => {
      return new Promise((res) => {
        res();
      });
    }
  };

  const journey = 'storageDocument';
  const documentNumber = 'some-document-number';

  beforeEach(() => {
    getUserReference.mockReturnValue({ type: 'save_user_reference' });
  });

  it('will call all methods needed to load the component', async () => {
    userReference.documentNumber = documentNumber;

    await userReference.loadData(store, journey);

    expect(getUserReference).toHaveBeenCalledWith(documentNumber);
  });

});

describe('When component did update service call fails', () => {

  it('should push history to forbidden page for unauthorised attempt when reference.unauthorised is true', () => {
    const mockPush = jest.fn();

    new UserReferencePage.WrappedComponent({
      match: {
        params: {
          documentNumber: 'GBR-23423-4234234'
        }
      },
      reference: { unauthorised: true },
      history: {
        push: mockPush
      },
      route: {
        title: 'Create a UK catch certificate - GOV.UK',
        header: 'Add your reference for this export',
        path: '/create-catch-certificate/:documentNumber/add-your-reference',
        previousUri: '/create-catch-certificate/catch-certificates',
        nextUri: '/create-catch-certificate/:documentNumber/add-exporter-details',
        journey: 'catchCertificate',
        journeyText: 'Catch Certificate'
      }
    }).componentDidUpdate();

    expect(mockPush).toHaveBeenCalledWith('/forbidden');
  });
});

describe('UserReference with errors', () => {

  let wrapper;

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  let mockPush;

  const props = {
    route: {
      title: 'Create a UK catch certificate - GOV.UK',
      header: 'commonAddYourReferenceForThisExport',
      path: '/create-catch-certificate/:documentNumber/add-your-reference',
      previousUri: '/create-catch-certificate/:documentNumber/progress',
      nextUri: '/create-catch-certificate/:documentNumber/add-exporter-details',
      saveAsDraftUri: '/create-catch-certificate/catch-certificates',
      journey: 'catchCertificate',
      journeyText: 'catch certificate'
    },
    match: {
      params: {
        documentNumber: 'GBR-23423-4234234'
      }
    }
  };

  const documentNumber = 'GBR-23423-4234234';
  const store = mockStore({
    reference: {
      userReference: 'My Reference',
      error: { userReference: 'error.userReference.string.pattern.base' },
    },
    confirmCopyDocument: {
      copyDocumentAcknowledged: false,
      voidDocumentConfirm: false,
    },
    errors: {},
  });

  beforeEach(() => {
    window.scrollTo = jest.fn();

    const history = createMemoryHistory({ initialEntries: [`/create-catch-certificate/${documentNumber}/add-your-reference`] });

    mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);
    wrapper = mount(
     <Provider store={store}>
        <Router history={history}>
          <Route path='/create-catch-certificate/:documentNumber/add-your-reference'>
            <UserReferencePage {...props}/>
          </Route>
        </Router>
      </Provider>
    );
  });

  it('should show errorIsland', () => {
    wrapper.find('input[name="userReference"]').simulate('blur', { target: { name: 'userReference', value: '<?!@##' } });
    wrapper.setState({selected: '/create-catch-certificate/catch-certificates'});
    wrapper.find('button#continue').simulate('click');
    expect(wrapper).toBeDefined();
    expect(wrapper.find('input[name="userReference"]').instance().value).toEqual('<?!@##');
    expect(wrapper.find('ErrorIsland li').at(0).text()).toEqual('Enter your reference as a combination of letters, numbers, hyphens, slashes and full stops');
  });
});

describe('UserReference back link Navigation for storage Notes Journey', () => {

  let wrapper;

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  let mockPush;

  const props = {
    route: {
      title: 'Create a UK storage document - GOV.UK',
      header: 'commonAddYourReferenceForThisExport',
      path: '/create-storage-document/:documentNumber/add-your-reference',
      previousUri: '/create-storage-document/:documentNumber/progress',
      nextUri: '/create-storage-document/:documentNumber/add-exporter-details',
      saveAsDraftUri: '/create-storage-document/catch-certificates',
      journey: 'storageNotes',
      journeyText: 'storageNotes'
    },
    match: {
      params: {
        documentNumber: 'GBR-23423-4234234'
      }
    }
  };

  const documentNumber = 'GBR-23423-4234234';
  const store = mockStore({
    reference: {},
    confirmCopyDocument: {},
    errors: {},
  });

  beforeEach(() => {
    window.scrollTo = jest.fn();

    const history = createMemoryHistory({ initialEntries: [`/create-storage-document/${documentNumber}/add-your-reference`] });

    mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);
    wrapper = mount(
     <Provider store={store}>
        <Router history={history}>
          <Route path='/create-storage-document/:documentNumber/add-your-reference'>
            <UserReferencePage {...props}/>
          </Route>
        </Router>
      </Provider>
    );
  });

  it('should redirect to the SD progress page when back link is clicked in Storage document journey', () => {
    wrapper.find(BackLink).simulate('click');

    expect(mockPush).toHaveBeenCalledTimes(1);
    expect(mockPush).toHaveBeenCalledWith('/create-storage-document/GBR-23423-4234234/progress');
  });

  it('should have a back to the progress page link', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper.find('a[href="/create-storage-document/GBR-23423-4234234/progress"]').exists()).toBeTruthy();
  });
});
