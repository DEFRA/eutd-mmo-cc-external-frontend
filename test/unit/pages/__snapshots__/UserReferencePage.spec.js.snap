// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When redirected from copy comfirm document page for a processing statement should take a snapshot of the whole page 1`] = `
<div>
  <div
    class="css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <a
            class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
            href="/create-catch-certificate/catch-certificates"
          >
            Back
          </a>
          <div
            class="notification-banner"
            role="alert"
            style="border: 0px;"
          >
            <div
              class="notification-banner__header"
            >
              <h2
                class="notification-banner__title"
              >
                Important
              </h2>
            </div>
            <div
              class="notification-banner__content"
            >
              <p
                class="notification-banner__heading"
              >
                This draft was created by copying a document that has now been voided. You are reminded that you must not use a processing statement or data for catches that have already been exported as this is a serious offence and may result in enforcement action being taken.
              </p>
            </div>
          </div>
          <h1
            class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
          >
            Add your reference for this export
          </h1>
        </div>
      </div>
      <form
        action="/orchestration/api/v1/userReference?c=/create-catch-certificate/undefined/add-your-reference&n=/create-catch-certificate/undefined/add-exporter-details"
        method="POST"
      >
        <input
          name="journey"
          type="hidden"
          value="processingStatement"
        />
        <input
          name="currentUri"
          type="hidden"
          value="/create-catch-certificate/:documentNumber/add-your-reference"
        />
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-n58wva-StyledColumn e4u1u2g0"
          >
            <label
              class="e1vcssft0 e1vcssft0 css-5tg97n-StyledLabel-StyledHoc-StyledHoc e4u1u2g0"
              for="userReference"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              >
                Your reference (optional)
              </span>
              <span
                class="e1vcssft0 css-1whl0n4-StyledHint-StyledHoc e4u1u2g0"
              >
                Enter a reference to help you identify this processing statement within the service. This reference is for your own use and will not appear on the final certificate. For example, you could choose a reference number from your organisation.
              </span>
              
              <input
                autocomplete="off"
                class="formControl e1vcssft0 css-4n7f6s-StyledInput-StyledHoc e4u1u2g0"
                id="userReference"
                name="userReference"
                type="text"
                value=""
              />
            </label>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
            formaction="/orchestration/api/v1/userReference/processingStatement/saveAsDraftLink"
            id="saveAsDraft"
            name="saveAsDraft"
            type="submit"
          >
            Save as Draft
          </button>
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
          >
            Save and continue
          </button>
        </div>
      </form>
      <a
        href="/create-catch-certificate/catch-certificates"
        id="backToProgress"
      >
        Back to Your Progress
        <span
          class="govuk-visually-hidden"
        >
          back to your progress page link
        </span>
      </a>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
